File: src/daemon/routes/evaluate.ts
/**
 * Author: Deepankar Das
 */

/**
 * Enforcer — /v1/evaluate Route Handler
 *
 * Accepts an ActionRequest from enforcement points, evaluates it
 * against the policy bundle, emits an audit event, and returns
 * the PolicyDecision.
 *
 * TDD Reference: Section 7.2
 */

import { v4 as uuidv4 } from "uuid";
import { ActionRequestSchema, type ActionRequest } from "../../../types/action.js";
import { type PolicyBundle, type PolicyDecision } from "../../../types/policy.js";
import { type AuditEvent } from "../../../types/audit-event.js";
import { evaluatePolicy } from "../../policy/engine.js";
import { buildGateFields } from "../../audit/validate.js";
import { type AuditBuffer } from "../../audit/buffer.js";
import { type ApprovalService } from "../../approval/service.js";

export interface EvaluateResult {
  status: number;
  body: PolicyDecision | { error: string; details?: string[] };
}

/**
 * Handle POST /v1/evaluate
 *
 * 1. Parse and validate the ActionRequest
 * 2. Check active approval scopes (auto-approve if matching)
 * 3. Evaluate policy
 * 4. If require_approval: create approval and wait (async)
 * 4. Return PolicyDecision
 */
export function handleEvaluate(
  rawBody: string,
  policyBundle: PolicyBundle,
  auditBuffer: AuditBuffer,
  approvalService?: ApprovalService,
): EvaluateResult {
  const evaluateStart = Date.now();

  // 1. Parse request body
  let parsed: unknown;
  try {
    parsed = JSON.parse(rawBody);
  } catch {
    return {
      status: 400,
      body: { error: "Invalid JSON" },
    };
  }

  // 2. Validate against ActionRequest schema
  const validation = ActionRequestSchema.safeParse(parsed);
  if (!validation.success) {
    return {
      status: 400,
      body: {
        error: "Invalid ActionRequest",
        details: validation.error.issues.map(i => `${i.path.join(".")}: ${i.message}`),
      },
    };
  }

  const request: ActionRequest = validation.data;

  // 2b. Check active approval scopes (auto-approve without re-prompting)
  if (approvalService) {
    const scopeMatch = approvalService.checkScope(request);
    if (scopeMatch) {
      // Auto-approved by matching scope — return allow decision
      const autoDecision: PolicyDecision = {
        request_id: request.request_id,
        decision: "allow",
        reason_code: "SCOPE_AUTO_APPROVED",
        reason_human: scopeMatch.rationale || "Auto-approved by matching scope.",
        policy_id: "scope.auto",
        policy_version: policyBundle.bundle_version,
        approval_required: false,
      };

      // Build and buffer audit event for scope match
      const autoGateFields = buildGateFields({
        actor: request.actor,
        session_id: request.actor.session_id,
        action: request.action,
        timestamp: request.timestamp,
        policy_detail: { policy_id: "scope.auto", policy_version: policyBundle.bundle_version },
        decision: "allow:SCOPE_AUTO_APPROVED",
        result: "pending",
      });

      const autoAuditEvent: AuditEvent = {
        event_id: uuidv4(),
        timestamp: request.timestamp,
        session_id: request.actor.session_id,
        correlation_id: request.request_id,
        ...autoGateFields,
        actor: { user_id: request.actor.user_id, agent_type: request.actor.agent_type, agent_instance: request.actor.agent_instance },
        environment: { workspace: request.environment.workspace, repo: request.environment.repo, branch: request.environment.branch, tier: request.environment.tier, deployment_mode: request.environment.deployment_mode },
        action: { type: request.action.type, attempted_action: request.action.attempted_action, observed_effect: "pending" },
        resource: { kind: request.resource.kind, path: request.resource.path, host: request.resource.host, value: request.resource.value, classification: request.resource.classification || [] },
        policy_detail: { policy_id: "scope.auto", policy_version: policyBundle.bundle_version, decision: "allow", reason_code: "SCOPE_AUTO_APPROVED", reason_human: "Auto-approved by matching scope." },
        approval: { status: "approved", approver_id: "scope_auto", rationale: scopeMatch.rationale, is_break_glass: false },
      };

      auditBuffer.bufferEvent(autoAuditEvent);
      return { status: 200, body: autoDecision };
    }
  }

  // 3. Evaluate policy
  const decision: PolicyDecision = evaluatePolicy(request, policyBundle);

  // 4. Build audit event
  const observedEffect = decision.decision === "allow"
    ? "pending"  // Will be updated by post_tool_call hook with actual result
    : decision.decision === "deny"
      ? "blocked"
      : "pending_approval";

  const decisionString = `${decision.decision}:${decision.reason_code}`;

  const gateFields = buildGateFields({
    actor: request.actor,
    session_id: request.actor.session_id,
    action: request.action,
    timestamp: request.timestamp,
    policy_detail: {
      policy_id: decision.policy_id,
      policy_version: decision.policy_version,
    },
    decision: decisionString,
    result: observedEffect,
  });

  const auditEvent: AuditEvent = {
    event_id: uuidv4(),
    timestamp: request.timestamp,
    session_id: request.actor.session_id,
    correlation_id: request.request_id,
    ...gateFields,
    actor: {
      user_id: request.actor.user_id,
      agent_type: request.actor.agent_type,
      agent_instance: request.actor.agent_instance,
    },
    environment: {
      workspace: request.environment.workspace,
      repo: request.environment.repo,
      branch: request.environment.branch,
      tier: request.environment.tier,
      deployment_mode: request.environment.deployment_mode,
    },
    action: {
      type: request.action.type,
      attempted_action: request.action.attempted_action,
      observed_effect: observedEffect,
    },
    resource: {
      kind: request.resource.kind,
      path: request.resource.path,
      host: request.resource.host,
      value: request.resource.value,
      classification: request.resource.classification || [],
    },
    policy_detail: {
      policy_id: decision.policy_id,
      policy_version: decision.policy_version,
      decision: decision.decision,
      reason_code: decision.reason_code,
      reason_human: decision.reason_human,
    },
  };

  // 5. Buffer audit event (async — never blocks the decision path)
  auditBuffer.bufferEvent(auditEvent);

  // 6. Log latency
  const evaluateLatencyMs = Date.now() - evaluateStart;
  if (evaluateLatencyMs > 50) {
    console.warn(`[EVALUATE] Slow policy decision: ${evaluateLatencyMs}ms for ${request.action.type} (target: p95 <50ms)`);
  }

  // 7. Return decision
  return {
    status: 200,
    body: decision,
  };
}

File: go/internal/daemon/routes/evaluate.go
/**
 * Author: Deepankar Das
 */

package routes

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/anthropics/enforcer/internal/approval"
	"github.com/anthropics/enforcer/internal/audit"
	"github.com/anthropics/enforcer/internal/policy"
	"github.com/anthropics/enforcer/internal/types"
)

// HandleEvaluate processes a policy evaluation request. It validates the
// action request, checks enforcement state, checks approval scopes,
// evaluates policy, builds an audit event, and buffers it.
// enforcementEnabled is passed from the daemon to avoid import cycles.
func HandleEvaluate(body []byte, bundle *types.PolicyBundle, buffer *audit.AuditBuffer, approvalSvc *approval.ApprovalService, enforcementEnabled bool) (int, interface{}) {
	// Parse request.
	var req types.ActionRequest
	if err := json.Unmarshal(body, &req); err != nil {
		return http.StatusBadRequest, map[string]string{
			"error":   "invalid_json",
			"message": fmt.Sprintf("Failed to parse action request: %s", err.Error()),
		}
	}

	// Validate.
	if errs := types.ValidateActionRequest(&req); len(errs) > 0 {
		return http.StatusBadRequest, map[string]interface{}{
			"error":            "validation_failed",
			"validation_errors": errs,
		}
	}

	// Check enforcement state.
	if !enforcementEnabled {
		return http.StatusOK, map[string]interface{}{
			"request_id":   req.RequestID,
			"decision":     "allow",
			"reason_code":  "ENFORCEMENT_DISABLED",
			"reason_human": "Enforcement is currently disabled. Action allowed.",
			"enforcement":  false,
		}
	}

	// Check approval scopes for pre-approved actions.
	if scopeDecision := approvalSvc.CheckScope(req); scopeDecision != nil {
		decision := types.PolicyDecision{
			RequestID:        req.RequestID,
			Decision:         types.DecisionAllow,
			ReasonCode:       "SCOPE_APPROVED",
			ReasonHuman:      fmt.Sprintf("Action pre-approved by scope (approval %s)", scopeDecision.ApprovalID),
			PolicyID:         "system.scope_approval",
			PolicyVersion:    bundle.BundleVersion,
			ApprovalRequired: false,
		}
		buildAndBufferEvent(req, decision, buffer)
		return http.StatusOK, decision
	}

	// Evaluate policy.
	decision := policy.EvaluatePolicy(req, *bundle)

	// If approval is required, create an approval request.
	if decision.Decision == types.DecisionRequireApproval {
		approvalReq, _ := approvalSvc.CreateApproval(req, decision)
		buildAndBufferEvent(req, decision, buffer)
		return http.StatusOK, map[string]interface{}{
			"request_id":        decision.RequestID,
			"decision":          string(decision.Decision),
			"reason_code":       decision.ReasonCode,
			"reason_human":      decision.ReasonHuman,
			"policy_id":         decision.PolicyID,
			"policy_version":    decision.PolicyVersion,
			"approval_required": true,
			"approval_id":       approvalReq.ApprovalID,
			"timeout_seconds":   approvalReq.TimeoutSeconds,
		}
	}

	buildAndBufferEvent(req, decision, buffer)
	return http.StatusOK, decision
}

// buildAndBufferEvent constructs an audit event from the request and decision,
// then buffers it for later flush.
func buildAndBufferEvent(req types.ActionRequest, decision types.PolicyDecision, buffer *audit.AuditBuffer) {
	now := time.Now().UTC().Format(time.RFC3339)

	gateFields := audit.BuildGateFields(
		req.Actor,
		req.Actor.SessionID,
		req.Action,
		now,
		decision.PolicyID,
		decision.PolicyVersion,
		string(decision.Decision),
		string(decision.Decision),
	)

	classifications := make([]string, len(req.Resource.Classification))
	for i, c := range req.Resource.Classification {
		classifications[i] = string(c)
	}

	event := types.AuditEvent{
		EventID:       fmt.Sprintf("evt_%s_%d", req.RequestID, time.Now().UnixNano()),
		Timestamp:     now,
		SessionID:     req.Actor.SessionID,
		CorrelationID: req.RequestID,

		Who:      gateFields["who"],
		What:     gateFields["what"],
		When:     gateFields["when"],
		Policy:   gateFields["policy"],
		Decision: gateFields["decision"],
		Result:   gateFields["result"],
	}

	event.Actor.UserID = req.Actor.UserID
	event.Actor.AgentType = req.Actor.AgentType
	event.Actor.AgentInstance = req.Actor.AgentInstance

	event.Environment.Workspace = req.Environment.Workspace
	event.Environment.Repo = req.Environment.Repo
	event.Environment.Branch = req.Environment.Branch
	event.Environment.Tier = req.Environment.Tier
	event.Environment.DeploymentMode = req.Environment.DeploymentMode

	event.Action.Type = string(req.Action.Type)
	event.Action.AttemptedAction = req.Action.AttemptedAction
	// Set effect based on the policy decision — the event is created before
	// tool execution, so we know the outcome from the decision.
	switch decision.Decision {
	case types.DecisionAllow, types.DecisionAllowDegraded, types.DecisionSimulate:
		event.Action.ObservedEffect = "executed"
	case types.DecisionDeny:
		event.Action.ObservedEffect = "blocked"
	case types.DecisionRequireApproval:
		event.Action.ObservedEffect = "pending_approval"
	default:
		event.Action.ObservedEffect = "executed"
	}

	event.Resource.Kind = string(req.Resource.Kind)
	event.Resource.Path = req.Resource.Path
	event.Resource.Host = req.Resource.Host
	event.Resource.Value = req.Resource.Value
	event.Resource.Classification = classifications

	event.PolicyDetail.PolicyID = decision.PolicyID
	event.PolicyDetail.PolicyVersion = decision.PolicyVersion
	event.PolicyDetail.Decision = string(decision.Decision)
	event.PolicyDetail.ReasonCode = decision.ReasonCode
	event.PolicyDetail.ReasonHuman = decision.ReasonHuman

	buffer.BufferEvent(event)
}
